subsys/bluetooth/host/smp.c

```diff
@@ -3181,34 +3182,40 @@ static uint8_t smp_pairing_rsp(struct bt_smp *smp, struct net_buf *buf)
 #endif /* CONFIG_BT_SMP_SC_PAIR_ONLY */
 	}
 
-	smp->local_dist &= SEND_KEYS_SC;
-	smp->remote_dist &= RECV_KEYS_SC;
+	if (IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY)) {
+		CODE_UNREACHABLE;
```

Comment:
unfortunately this makes the compiler complain because the function exits with `return sc_send_public_key(smp)` and `sc_send_public_key()` uses the `sc_public_key` NULL in a memcpy

Reply:
I'm not sure I understand. The code below `const struct bt_conn_auth_cb *smp_auth_cb = latch_auth_cb(smp);` in the above suggestion will remain unchanged, so why would it not work? 

You said that `SMP_FLAG_SC` will be false if `CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY`, so it will never enter the `else` clause when `CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY=y`.

Reply:
it seems that is a compiler thing. When having `CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY`  activated and using `if (IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY))` the compiler will know for sure that code will branch in that if. But if we use
```
if (!atomic_test_bit(smp->flags, SMP_FLAG_SC))  {
} else {
}
```
the compiler will not know for sure which branch will be taken and will complain about the code in the else

Reply:
Right. 

I don't have a good solution for that right now, but I'm 100% sure that is it possible to avoid using `CODE_UNREACHABLE`. I don't think `CODE_UNREACHABLE` should be used, but won't block this PR if others approve

Reply:
the solution to completely avoid compiler warning and usage of `CODE_UNREACHABLE` when `CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY` is true would be to do replace `if (!atomic_test_bit(smp->flags, SMP_FLAG_SC))` with `if (IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY) || !atomic_test_bit(smp->flags, SMP_FLAG_SC))`

Reply:
That sounds like a decent solution to me - What do you think?

Reply:
I would do it like that and also add a comment in code on the reason of doing it like that, to have some context for people that will ask themselves why we did that since `IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY)` always implies `SMP_FLAG_SC` being false and it will seem an useless OR op

Reply:
Do not use `CODE_UNREACHABLE`! It should be named `INVOKE_UNDEFINED_BEHAVIOR` because that's exactly what it does. It is dangerous and should only be used in rare cases for optimization.

Reply:
> I would do it like that and also add a comment in code on the reason of doing it like that, to have some context for people that will ask themselves why we did that since IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY) always implies SMP_FLAG_SC being false and it will seem an useless OR op

A comment would be nice :)

Reply:
updated